crates/block-producer/src/block_sync_client.rs

```diff
@@ -0,0 +1,461 @@
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+async fn handle_local_block(
+    client: &mut BlockSyncClient,
+    l: gw_types::packed::LocalBlock,
+) -> Result<(), anyhow::Error> {
+    let block_hash = l.block().hash();
+    let block_number = l.block().raw().number().unpack();
+    log::info!(
+        "received block {block_number} {}",
+        ckb_types::H256::from(block_hash),
+    );
+    let store_tx = client.store.begin_transaction();
+    let store_block_hash = store_tx.get_block_hash_by_number(block_number)?;
+    if let Some(store_block_hash) = store_block_hash {
+        if store_block_hash != block_hash.into() {
+            log::info!("revert to {}", block_number - 1);
+            revert(client, &store_tx, block_number - 1).await?;
+            store_tx.commit()?;
+        } else {
+            log::info!("block already known");
+            return Ok(());
+        }
+    }
+    {
+        log::info!("update local block");
+        let mut chain = client.chain.lock().await;
+        block_in_place(|| {
+            let store_tx = client.store.begin_transaction();
+            chain.update_local(
+                &store_tx,
+                l.block(),
+                l.deposit_info_vec(),
+                l.deposit_asset_scripts().into_iter().collect(),
+                l.withdrawals().into_iter().collect(),
+                l.post_global_state(),
+            )?;
+            chain.calculate_and_store_finalized_custodians(&store_tx, block_number)?;
+            store_tx.commit()?;
+            anyhow::Ok(())
+        })?;
+    }
+    notify_new_tip(client, false).await?;
+    Ok(())
+}
+
+pub struct P2PStream {
+    id: SessionId,
+    control: ServiceAsyncControl,
+    read_part: Option<SubstreamReadPart>,
+    decoder: StreamDecoder,
+}
+
+impl P2PStream {
+    /// After calling this, you can only receive from the returned stream, and
+    /// self can only be used for disconnecting. (This is for receiving from
+    /// another task.)
+    fn take_receiver(&mut self) -> Self {
+        Self {
+            id: self.id,
+            control: self.control.clone(),
+            read_part: self.read_part.take(),
+            decoder: core::mem::take(&mut self.decoder),
+        }
+    }
+
+    async fn recv(&mut self) -> Result<Option<Bytes>> {
+        let receiver = self.read_part.as_mut().context("stream is taken")?;
+        Ok(if let Some(msg) = receiver.try_next().await? {
+            // Decompress message.
+            Some(self.decoder.decode(&msg)?.into())
+        } else {
+            None
+        })
+    }
+
+    async fn send(&mut self, msg: Bytes) -> Result<()> {
+        self.control
+            .send_message_to(self.id, P2P_SYNC_PROTOCOL, msg)
+            .await?;
+        Ok(())
+    }
+
+    async fn disconnect(&mut self) -> Result<()> {
+        self.control.disconnect(self.id).await?;
+        Ok(())
+    }
+}
+
+// XXX: would unbounded channel leak memory?
```

Comment:
P2P allowlist?

Reply:
Dead streams will be accumulated in this channel if we are syncing from l1 and in the mean time re-connecting.

Let me change this to `Mutex` or `Watch` or something else.